### PR TITLE
gh-99240: Fix double-free bug in Argument Clinic str_converter generated code

### PR DESCRIPTION
**Pull request by colorfulappl**
Fix double-free bug mentioned at https://github.com/python/cpython/issues/99240,
by moving memory clean up out of "exit" label.

* Issue: gh-99240

Automerge-Triggered-By: GH:erlend-aasland

### CODE REVIEW

File: Tools/clinic/clinic.py
Comment (gpshead) [inline Tools/clinic/clinic.py]: these two lines can use a walrus to become `if post_operations := self.post_operations():`

**Review by gpshead — APPROVED** [review]
overall this looks good, only a minor Python code modernization suggestion.

I do think your second bullet from the bug "If function _PyArg_ParseStack parses failed, assign all the parsed arguments to "NULL" after they are freed, this should be done in _PyArg_ParseStack." is also worth doing to make bugs like this less likely to occur. That can be done in its own PR.

**Comment by gpshead** [comment]
Of particular note, since the generated files check in CI passed, does that mean we don't have any actual standard library argument clinic uses of this str conversion via codec functionality yet?

**Comment by colorfulappl** [comment]
> does that mean we don't have any actual standard library argument clinic uses of this str conversion via codec functionality yet?

Yes, I have searched this in CPython source code and didn't find any usage of argument clinic str conversion with "encoding" parameter setting.

And test of this functionality is added in https://github.com/python/cpython/pull/96178, which is not merged yet.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: The distinction between `post_operations` and `cleanup` is not immediately obvious from the name only. Perhaps `post_parsing` could be an improvement?

**Comment by erlend-aasland** [comment]
Also, my preference, as stated in #96178, would be to _first_ merge that PR (with a _minimal_ test suite), and then for each bugfix PR (such as this one) add both the bugfix and the complementing test. So, I'm suggesting putting this on hold until #96178 is merged.

**Comment by colorfulappl** [comment]
Added Argument Clinic functional test (#96002).

File: Modules/clinic/_testclinic.c.h
Comment (arhadthedev) [inline Modules/clinic/_testclinic.c.h]: Testing for `NULL` is not required. From https://docs.python.org/3/c-api/memory.html#c.PyMem_Free:

> If p is NULL, no operation is performed.

So code like this can be reduced:

```diff
-    if (a) {
        PyMem_FREE(a);
-    }
```

*Edit: Oh, nevermind, now I see in clinic.py that [generation of this line](https://github.com/python/cpython/pull/99241/files#diff-4c6b55ec7d80a3d60f100ce230eb879d9acbe2e558e01d318960e6197d7c47b3R3443) was introduced before your PR.*

File: Tools/clinic/clinic.py
Comment (arhadthedev) [inline Tools/clinic/clinic.py:3443]: Oh, nevermind, now I see it's out of scope of your PR.

File: Misc/NEWS.d/next/Library/2022-11-08-15-54-43.gh-issue-99240.MhYwcz.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2022-11-08-15-54-43.gh-issue-99240.MhYwcz.rst]: How about this?

```suggestion
Fix double-free bug in Argument Clinic ``str_converter`` by
extracting memory clean up to a new ``post_parsing`` section.
```

File: Modules/_testclinic.c
Comment (erlend-aasland) [inline Modules/_testclinic.c]: I don't think we need to repeat what was said in the issue, so this can go, IMO.

```suggestion
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Nit:

```suggestion
        Return a string containing this code indented at column 0.
        If no operation is necessary, return an empty string.
```

**Review by erlend-aasland — COMMENTED** [review]
LGTM with minor nitpicks :)

File: Modules/clinic/_testclinic.c.h
Comment (colorfulappl) [reply Modules/clinic/_testclinic.c.h]: > in clinic.py that [generation of this line](https://github.com/python/cpython/pull/99241/files#diff-4c6b55ec7d80a3d60f100ce230eb879d9acbe2e558e01d318960e6197d7c47b3R3443) was introduced before your PR.

Should I fix this problem in current PR?

Comment (erlend-aasland) [reply Modules/clinic/_testclinic.c.h]: > Should I fix this problem in current PR?

I'm fine with including that enhancement in this PR.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: One last nit:

```suggestion
            data.post_parsing.append('/* Post parse cleanup for ' + name + ' */\n' + post_parsing.rstrip() + '\n')
```

**Review by erlend-aasland — APPROVED** [review]
Thanks!

**Comment by erlend-aasland** [comment]
> I do think your second bullet from the bug "If function _PyArg_ParseStack parses failed, assign all the parsed arguments to "NULL" after they are freed, this should be done in _PyArg_ParseStack." is also worth doing to make bugs like this less likely to occur. That can be done in its own PR.

@colorfulappl, are you up for fixing `_PyArg_ParseStack` while you're at it?

**Comment by colorfulappl** [comment]
> are you up for fixing `_PyArg_ParseStack` while you're at it?

I have not started yet, but I am willing to have a try. :)

**Pull request merged**
